File: Govorilka/Services/TextCleanerService.swift
import Foundation

/// Service for cleaning transcribed text from filler words (слова-паразиты)
final class TextCleanerService {
    static let shared = TextCleanerService()

    /// Patterns for filler words removal (regex pattern, replacement)
    private let fillerPatterns: [(regex: NSRegularExpression, replacement: String)]

    private init() {
        // Define patterns for Russian filler words
        let patterns: [(String, String)] = [
            // Междометия и звуки (отдельно стоящие)
            (#"\b[эе]+\b,?\s*"#, ""),           // э, ээ, эээ
            (#"\b[мm]+\b,?\s*"#, ""),            // м, мм, ммм
            (#"\bугу\b,?\s*"#, ""),
            (#"\bага\b,?\s*"#, ""),

            // Вводные слова (более длинные паттерны сначала)
            (#"\bэто самое\b,?\s*"#, ""),
            (#"\bтак сказать\b,?\s*"#, ""),
            (#"\bв общем-то\b,?\s*"#, ""),
            (#"\bв общем\b,?\s*"#, ""),
            (#"\bв принципе\b,?\s*"#, ""),
            (#"\bну вот\b,?\s*"#, ""),
            (#"\bкак бы\b,?\s*"#, ""),
            (#"\bну\b,?\s*"#, ""),
            (#"\bвот\b,?\s*"#, ""),
            (#"\bтипа\b,?\s*"#, ""),
            (#"\bтипо\b,?\s*"#, ""),
            (#"\bкороче\b,?\s*"#, ""),
            (#"\bзначит\b,?\s*"#, ""),
            (#"\bсобственно\b,?\s*"#, ""),
            (#"\bслушай\b,?\s*"#, ""),
            (#"\bсмотри\b,?\s*"#, ""),
        ]

        // Compile regex patterns
        fillerPatterns = patterns.compactMap { pattern, replacement in
            guard let regex = try? NSRegularExpression(
                pattern: pattern,
                options: [.caseInsensitive]
            ) else {
                print("[TextCleanerService] Failed to compile pattern: \(pattern)")
                return nil
            }
            return (regex, replacement)
        }

        print("[TextCleanerService] Initialized with \(fillerPatterns.count) patterns")
    }

    /// Clean text by removing filler words
    /// - Parameter text: Original transcribed text
    /// - Returns: Cleaned text
    func clean(_ text: String) -> String {
        guard !text.isEmpty else { return text }

        var result = text

        // Apply all filler patterns
        for (regex, replacement) in fillerPatterns {
            result = regex.stringByReplacingMatches(
                in: result,
                options: [],
                range: NSRange(result.startIndex..., in: result),
                withTemplate: replacement
            )
        }

        // Normalize whitespace (collapse multiple spaces)
        result = result.replacingOccurrences(
            of: #"\s{2,}"#,
            with: " ",
            options: .regularExpression
        )

        // Trim leading/trailing whitespace
        result = result.trimmingCharacters(in: .whitespaces)

        // Capitalize first letter if text starts with lowercase
        if let first = result.first, first.isLowercase {
            result = first.uppercased() + result.dropFirst()
        }

        return result
    }
}

File: govorilka-mobile/src/services/TextCleanerService.ts
// Filler words and patterns to remove (Russian)
const FILLER_PATTERNS: RegExp[] = [
  /\b(ну|эм|эээ|ааа|ммм)\b/gi,
  /\b(типа|как бы|в общем|короче|значит)\b/gi,
  /\b(это самое|так сказать|в принципе)\b/gi,
  /\b(вот|ну вот|ну типа)\b/gi,
];

// Clean up extra spaces and punctuation
const CLEANUP_PATTERNS: [RegExp, string][] = [
  [/\s{2,}/g, ' '],           // Multiple spaces → single
  [/\s+([.,!?;:])/g, '$1'],    // Space before punctuation
  [/^\s+/gm, ''],              // Leading spaces on lines
  [/\s+$/gm, ''],              // Trailing spaces on lines
];

class TextCleanerServiceClass {
  clean(text: string): string {
    let result = text;

    // Remove filler words
    for (const pattern of FILLER_PATTERNS) {
      result = result.replace(pattern, '');
    }

    // Clean up formatting
    for (const [pattern, replacement] of CLEANUP_PATTERNS) {
      result = result.replace(pattern, replacement);
    }

    // Capitalize first letter after period
    result = result.replace(/\.\s+([a-zа-яё])/g, (_, char) => `. ${char.toUpperCase()}`);

    // Capitalize first letter of text
    if (result.length > 0) {
      result = result.charAt(0).toUpperCase() + result.slice(1);
    }

    return result.trim();
  }
}

export const TextCleanerService = new TextCleanerServiceClass();
